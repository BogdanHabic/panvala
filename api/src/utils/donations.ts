const { Donation } = require('../models');

// Base transaction info -- all required
export interface IDonationTx {
  txHash: string;
  metadataHash: string;
  sender: string;
  donor: string;
  tokens: string;
}

// Donation data that is public
export interface IPublicDonation extends IDonationTx {
  metadataVersion?: string;
  memo?: string;
  usdValueCents?: string;
  ethValue?: string;
  pledgeMonthlyUSDCents?: number;
  pledgeTerm?: number;
}

export interface IDonation extends IPublicDonation {
  firstName?: string;
  lastName?: string;
  email?: string;
  company?: string;
  fundraiser?: string;
}

export function addDonation(donation: IDonation) {
  return Donation.create(donation);
}

const publicFields = [
  'txHash',
  'metadataHash',
  'sender',
  'donor',
  'tokens',
  'metadataVersion',
  'memo',
  'usdValueCents',
  'ethValue',
  'pledgeMonthlyUSDCents',
  'pledgeTerm',
];

export function getPublicDonations(): Promise<IPublicDonation[]> {
  return Donation.findAll({
    attributes: publicFields,
  });
}

export function getDonationsForFundraiser(fundraiser: string): Promise<IPublicDonation[]> {
  return Donation.findAll({
    attributes: publicFields,
    where: {
      fundraiser,
    },
  });
}
