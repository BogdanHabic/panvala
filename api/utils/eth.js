const ethers = require('ethers');
const config = require('./config');
const { rpcEndpoint } = config;
const { gatekeeperAddress, tokenCapacitorAddress } = config.contracts;

const {
  contractABIs: { Gatekeeper, ParameterStore, TokenCapacitor, BasicToken },
} = require('../../packages/panvala-utils');

/**
 * Check connection
 */
function checkConnection() {
  // console.log('Connecting to', rpcEndpoint);
  const provider = new ethers.providers.JsonRpcProvider(rpcEndpoint);

  return provider.getBlockNumber();
}

async function getContracts() {
  const provider = new ethers.providers.JsonRpcProvider(rpcEndpoint);
  const signer = provider.getSigner();
  const gatekeeper = new ethers.Contract(gatekeeperAddress, Gatekeeper.abi, signer);
  const tokenCapacitor = new ethers.Contract(tokenCapacitorAddress, TokenCapacitor.abi, signer);

  const parameterStoreAddress = await gatekeeper.functions.parameters();
  const parameterStore = new ethers.Contract(parameterStoreAddress, ParameterStore.abi, signer);

  return {
    gatekeeper,
    parameterStore,
    tokenCapacitor,
  };
}

module.exports = {
  checkConnection,
  getContracts,
};
