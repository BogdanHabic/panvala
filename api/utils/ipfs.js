const IPFS = require('ipfs-http-client');

const ipfsHost = process.env.IPFS_HOST || 'ipfs.infura.io';
const ipfsPort = process.env.IPFS_PORT || 5001;

const ipfs = new IPFS({ host: ipfsHost, port: ipfsPort, protocol: 'https' });

/**
 * Get a file from IPFS
 * Options:
 *   {Boolean} json Whether to parse as JSON
 * @param {String} multihash
 * @param {Object} options
 */
function get(multihash, options) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => {
      reject(new Error('ipfs request timed out'));
    }, 7000);

    ipfs.cat(multihash).then(response => {
      clearTimeout(timer);

      if (options.json) {
        try {
          const data = JSON.parse(response);
          resolve(data);
        } catch (error) {
          console.log(`ERROR: while JSON.parse result: ${error.message}`);
          throw error;
        }
      }
    });
  });
}

module.exports = {
  get,
};
