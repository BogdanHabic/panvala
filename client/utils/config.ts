import getConfig from 'next/config';

// Defaults are a workaround for https://github.com/zeit/next.js/issues/4024
const { publicRuntimeConfig = {} } = getConfig() || {};

export function loadConfig() {
  return publicRuntimeConfig;
}
