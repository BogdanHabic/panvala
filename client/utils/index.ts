// Re-export Panvala utils to be used throughout
import { default as panvala_utils } from '../../packages/panvala-utils';
export { panvala_utils };
