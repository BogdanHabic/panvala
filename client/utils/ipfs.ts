const IPFS = require('ipfs-mini');
const isIPFS = require('is-ipfs')

const ipfs: any = new IPFS({ host: 'ipfs.infura.io', port: 5001, protocol: 'https' });

export function ipfsCheckMultihash(multihash: string): boolean | Error {
  if (isIPFS.multihash(multihash)) {
    return true;
  }
  throw new Error('invalid multihash');
}

export async function ipfsGetData(multihash: string) {
  if (ipfsCheckMultihash(multihash)) {
    return new Promise((resolve, reject) => {
      ipfs.catJSON(multihash, (err: any, result: string) => {
        if (err) reject(new Error(err));
        resolve(result);
      });
    });
  }
}

export async function ipfsAddObject(obj: any): Promise<string> {
  const CID: string = await new Promise((resolve, reject) => {
    ipfs.addJSON(obj, (err: any, result: string) => {
      if (err) reject(new Error(err));
      resolve(result);
    });
  });
  console.log('CID:', CID);
  return CID;
}
