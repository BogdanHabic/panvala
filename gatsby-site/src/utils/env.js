export const Environment = {
  staging: 'staging',
  development: 'development',
  production: 'production',
};

/**
 * Determine the environment (production, staging, development) using the URL
 * TODO: use environment variables
 */
export function getEnvironment() {
  const urlRoute = window.location.href;

  const isStaging = urlRoute.includes('staging/donate') || urlRoute.includes('develop.panvala.com');

  const environment = isStaging
    ? Environment.staging
    : urlRoute.includes('localhost')
    ? Environment.development
    : Environment.production;

  return environment;
}
