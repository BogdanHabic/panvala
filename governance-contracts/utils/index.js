const ethers = require('ethers');
const ethUtils = require('ethereumjs-util');
const BigNumber = require('bignumber.js');


const { defaultAbiCoder: abiCoder } = ethers.utils;

/**
 * ABI-encode a value to pass to a contract function
 * @param {*} type
 * @param {*} value
 */
function abiEncode(type, value) {
  return abiCoder.encode([type], [value]);
}

function asBytes(string) {
  return ethUtils.toBuffer(string);
}

/**
 * stripHexPrefix
 * Remove '0x' from the beginning of a string, if present
 * @param {String} value
 * @return String
 */
function stripHexPrefix(value) {
  // assume string
  const stripped = value.startsWith('0x') ? value.substring(2) : value;
  return stripped;
}

/**
 * bytesAsString
 * @param {String} bytes Hex-encoded string returned from a contract
 * @return UTF-8 encoded string
 */
function bytesAsString(bytes) {
  const stripped = stripHexPrefix(bytes);
  const decoded = Buffer.from(stripHexPrefix(stripped), 'hex');
  return decoded.toString();
}

/**
 * zeroHash
 * @dev Return a 32-byte value of all zeros
 */
function zeroHash() {
  return ethUtils.zeros(32);
}

const ten = new BigNumber(10);
const panDecimals = new BigNumber(18);

/**
 * Convert the amount to PAN base units (string)
 * @param {number|string} tokens can be fractional
 * @return {string}
 */
function toPanBase(tokens) {
  const factor = ten.pow(panDecimals);
  const baseUnits = factor.times(new BigNumber(tokens));
  return baseUnits.toFixed();
}

/**
 *
 * @param {ethers.utils.BigNumberish} baseValue
 * @return {string}
 */
function fromPanBase(baseValue) {
  const base = new BigNumber(baseValue.toString());
  const factor = ten.pow(panDecimals);
  const tokens = base.div(factor);
  return tokens.toFixed();
}

module.exports = {
  BN: ethUtils.BN,
  abiEncode,
  asBytes,
  stripHexPrefix,
  bytesAsString,
  zeroHash,
  sha256: ethUtils.sha256,
  toPanBase,
  fromPanBase,
};
