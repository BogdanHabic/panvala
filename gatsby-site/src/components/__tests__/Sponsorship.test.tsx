import * as React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import Sponsorship from '../Sponsorship';
import { theme } from '../../utils/theme';

describe('Donation', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = render(
      <ThemeProvider theme={theme}>
        <Sponsorship />
      </ThemeProvider>
    );
  });

  test('should render component', () => {
    // Get the element
    const instance = container.getByTestId('sponsorship-container');

    // Snapshot
    expect(instance).toMatchSnapshot();
  });
});
