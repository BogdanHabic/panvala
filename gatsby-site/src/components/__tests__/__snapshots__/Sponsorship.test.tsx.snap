// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Donation should render component 1`] = `
<div
  class="sc-bwzfXH bkbqwe"
  data-testid="sponsorship-container"
>
  <form
    class="w-80-l w-90-m w-100 center"
    name="sponsorship-pledge"
  >
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
        required=""
      >
        Company
        <b
          class="red f7"
        >
           *
        </b>
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <input
      class="sc-bdVaJa kGEDHa input-reset b--black-10"
      id="pledge-company"
      label="Company"
      name="company"
      placeholder="Enter the name of your company"
      required=""
      type="text"
      value=""
    />
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
      >
        First Name
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <input
      class="sc-bdVaJa kGEDHa input-reset b--black-10"
      id="pledge-first-name"
      label="First Name"
      name="firstName"
      placeholder="Enter your first name"
      type="text"
      value=""
    />
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
      >
        Last Name
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <input
      class="sc-bdVaJa kGEDHa input-reset b--black-10"
      id="pledge-last-name"
      label="Last Name"
      name="lastName"
      placeholder="Enter your last name"
      type="text"
      value=""
    />
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
        required=""
      >
        Email
        <b
          class="red f7"
        >
           *
        </b>
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <input
      class="sc-bdVaJa kGEDHa input-reset b--black-10"
      id="pledge-email"
      label="Email"
      name="email"
      placeholder="Enter your email address"
      required=""
      type="text"
      value=""
    />
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
        required=""
      >
        Monthly Pledge Amount (USD)
        <b
          class="red f7"
        >
           *
        </b>
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <input
      class="sc-bdVaJa kGEDHa input-reset b--black-10"
      id="pledge-amount"
      label="Monthly Pledge Amount (USD)"
      min="0"
      name="pledgeAmount"
      placeholder="Enter your pledge amount"
      required=""
      step="1"
      type="number"
      value=""
    />
    <div
      class="tl mt4"
    >
      <label
        class="ma0 f6 mb3 black-40"
        required=""
      >
        How many months of your pledge will you prepay today?
        <b
          class="red f7"
        >
           *
        </b>
      </label>
    </div>
    <div
      class="red tl pt2"
    />
    <select
      class="f6 input-reset b--black-10 pv3 ph2 db center w-100 br3 mt2 bg-white black-50"
      id="pledge-duration-select"
      name="pledgeDuration"
      required=""
    >
      <option
        value="0"
      >
        Select the amount of months you would like to prepay for
      </option>
      <option
        value="1"
      >
        1 month
      </option>
      <option
        value="3"
      >
        3 months
      </option>
      <option
        value="6"
      >
        6 months
      </option>
      <option
        value="12"
      >
        12 months
      </option>
    </select>
    <img
      alt=""
      class="fr mr2 o-50"
      src="test-file-stub"
      style="margin-top: -35px;"
    />
    <div>
      <input
        class="f6 link dim bn br-pill pv3 ph4 white bg-teal fw7 mt4"
        name="submit"
        type="submit"
        value="Sponsor"
      />
    </div>
  </form>
</div>
`;
